src/main/java/org/opensearch/knn/index/codec/KNN990Codec/NativeEngines990KnnVectorsWriter.java

```diff
@@ -146,4 +132,101 @@ public long ramBytesUsed() {
+            case BYTE:
+                ByteVectorValues mergedBytes = MergedVectorValues.mergeByteVectorValues(fieldInfo, mergeState);
+                return (KNNVectorValues<T>) KNNVectorValuesFactory.getVectorValues(vectorDataType, mergedBytes);
+            default:
+                throw new IllegalStateException("Unsupported vector encoding [" + fieldInfo.getVectorEncoding() + "]");
+        }
+    }
+
+    /**
+     * Functional interface representing an operation that indexes the provided {@link KNNVectorValues}.
+     *
+     * @param <T> The type of vectors being processed.
+     */
+    @FunctionalInterface
+    private interface IndexOperation<T> {
+        void buildAndWrite(NativeIndexWriter writer, KNNVectorValues<T> knnVectorValues) throws IOException;
+    }
+
+    /**
+     * Functional interface representing a method that retrieves {@link KNNVectorValues} based on
+     * the vector data type, field information, and the merge state.
+     *
+     * @param <DataType>   The type of the data representing the vector (e.g., {@link VectorDataType}).
+     * @param <FieldInfo>  The metadata about the field.
+     * @param <MergeState> The state of the merge operation.
+     * @param <Result>     The result of the retrieval, typically {@link KNNVectorValues}.
+     */
+    @FunctionalInterface
+    private interface VectorValuesRetriever<DataType, FieldInfo, MergeState, Result> {
+        Result apply(DataType vectorDataType, FieldInfo fieldInfo, MergeState mergeState) throws IOException;
+    }
+
+    /**
+     * Unified method for processing a field during either the indexing or merge operation. This method retrieves vector values
+     * based on the provided vector data type and applies the specified index operation, potentially including quantization if needed.
+     *
+     * @param fieldInfo              The {@link FieldInfo} object containing metadata about the field.
+     * @param vectorValuesRetriever  A functional interface that retrieves {@link KNNVectorValues} based on the vector data type,
+     *                                field information, and additional context (e.g., merge state or field writer).
+     * @param indexOperation         A functional interface that performs the indexing operation using the retrieved
+     *                                {@link KNNVectorValues}.
+     * @param context                The additional context required for retrieving the vector values (e.g., {@link MergeState} or {@link NativeEngineFieldVectorsWriter}).
+     * @param <T>                    The type of vectors being processed.
+     * @param <C>                    The type of the context needed for retrieving the vector values.
+     * @throws IOException If an I/O error occurs during the processing.
+     */
+    private <T, C> void trainAndIndex(
+        final FieldInfo fieldInfo,
+        final VectorValuesRetriever<VectorDataType, FieldInfo, C, KNNVectorValues<T>> vectorValuesRetriever,
```

Comment:
This is really abuse of generics. Just because 2 functions are returning same thing doesn't mean that we should use generic to make them same. If you use some overloaded functions it could have been easy to read, understand and extend.

Reply:
IMO , it's looks fine.

Reply:
Would like to know the thoughts of other maintainers here and sorry it doesn't follow the KISS which you are trying to follow in other parts of the code.

cc: @jmazanec15 , @heemin32

Reply:
Well using generics and Functional Inteeface is also part of KISS. I would like to read any literature  where it says functional inteface is not following KISS.

Reply:
What I mean to say is functional Interface will be following KISS but this function doesn't. I don't know from where you get this idea I am saying Functional interface are not part of KISS.

Reply:
As you suggested Writing total 6 functions for flush and merge ...that's where we have generics and functional interface to solve the problem of writing 6 functions vs 1.

Reply:
This felt like a bit of an over-engineering IMO. Its creating 2 functional interfaces with little advantage and little possibility of reuse. Having said that there isn't a performance disadvantage here but I would still encourage you to simplify this in terms of readability considering the code base is not functional

Reply:
Is there any way we can avoid the Context arguement? I think everything else looks good, but the context is a bit hard to follow. Can VectorValuesRetriever just be configured from calling code and avoid context?

Reply:
I changed it to have more meaningful name and added more Java Docs around that to have better understanding , if In case we need there is more understanding issue , I can come back and change in future.